### PR TITLE
Fix Pydantic Unions

### LINKED ISSUE #1230: Pydantic conversion fails for Union types
If you include a field in a Pydantic object with a `Union[A, B]` type, where A and B are Pydantic objects, then conversion fails with `AttributeError: 'StrawberryUnion' object has no attribute '_type_definition'`. We need to handle the branches of the union separately to get the right behavior here.

```
class BranchA(pydantic.BaseModel):
    field_a: str

class BranchB(pydantic.BaseModel):
    field_b: int

class User(pydantic.BaseModel):
    age: int
    union_field: Union[BranchA, BranchB]

@strawberry.experimental.pydantic.type(BranchA, fields=["field_a"])
class BranchAType:
    pass

@strawberry.experimental.pydantic.type(BranchB, fields=["field_b"])
class BranchBType:
    pass

@strawberry.experimental.pydantic.type(User, fields=["age", "union_field"])
class UserType:
    pass

origin_user = User(age=1, union_field=BranchA(field_a="abc"))
user = UserType.from_pydantic(origin_user)
# raises AttributeError, should return UserType(age=1, union_field=BranchAType(field_a="abc"))
```

### PR DESCRIPTION
**Pull request by Matt343**
## Description

In order to convert Union types from Pydantic models into Strawberry models, we need to check which branch of the Union we have and proceed with the conversion according to that type definition. I had to add a `_pydantic_type` field to our `StrawberryType` instances that are created through the converter so that we can get a reference to the original type of each branch for comparison. This seems a little hacky, but I couldn't think of a better way.

This fix only works for Unions of object types, but that should be fine as GraphQL doesn't support other types of unions.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Closes #1230 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Fixes a bug in the Pydantic conversion code around `Union` values.



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Matt Allen for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by patrick91** [comment]
@Matt343 thanks! I'll try to review this over the weekend ^^

File: strawberry/experimental/pydantic/object_type.py [hunk 1]
```diff
@@ -144,6 +144,7 @@ def wrap(cls):
         )
 
         model._strawberry_type = cls  # type: ignore
+        cls._pydantic_type = model  # type: ignore
```
Comment (patrick91) [inline strawberry/experimental/pydantic/object_type.py:147]: this is a patter we want to move away from :)

I think we can keep this for now though, we still need to do some refactoring anyway ^^

**Review by patrick91 — APPROVED** [review]
Looks good! I've only added a note, but it is nothing to worry about for now, we'll deal with it when we finalise some refactoring and when we tackle this issue: https://github.com/strawberry-graphql/strawberry/issues/1192

**Pull request merged**